### PR TITLE
Buffer for HandleGetFile's LazyChunkReader

### PR DESCRIPTION
**Pull request by janos**
Related to: https://github.com/ethersphere/go-ethereum/pull/638
More information thanks to @nonsense: https://hackmd.io/QDyOyUQ3SNGD2JPjhWvdTA

TODO: Test performance gains and correctness using a custom Retrieve function.

### CODE REVIEW

File: swarm/api/http/server.go [hunk 1]
```diff
@@ -975,7 +976,38 @@ func (s *Server) HandleGetFile(w http.ResponseWriter, r *Request) {
...
 	w.Header().Set("Content-Type", contentType)
-	http.ServeContent(w, &r.Request, "", time.Now(), reader)
+	http.ServeContent(w, &r.Request, "", time.Now(), newBufferedReadSeeker(reader, getFileBufferSize))
+}
+
+// The size of buffer used for bufio.Reader on LazyChunkReader passed to
+// http.ServeContent in HandleGetFile.
+// Warning: This value influences the number of chunk requests and chunker join goroutines
+// per file request.
+// Recommended value is 4 times the io.Copy default buffer value which is 32kB.
```
Comment (nonsense) [inline swarm/api/http/server.go:986]: Why just 4*? Based on the branching factor we have, I think we can go much higher, like 32* or 64*, no?

Comment (janos) [reply swarm/api/http/server.go:986]: While testing on open cluster, raising over 4 did not improve download speed. It was saturated after 4 (32 goroutines), with 100 download was practically impossible as many chunk requests were timing out.

Comment (nonsense) [reply swarm/api/http/server.go:986]: Hmm... ok.

Comment (nonsense) [reply swarm/api/http/server.go:986]: This feels quite inefficient... if our requests are for single chunks, I don't think we will manage to make it much more efficient. We should probably batch the requests for chunks, similarly to the Syncing protocol.

This is just my intuition, I might be completely wrong here.

Comment (janos) [reply swarm/api/http/server.go:986]: Yes, some sort of concurrent chunk retrieval limit is required.

**Comment by janos** [comment]
@zelig this pr is ready for review to test it locally or on open cluster. It needs tests, and they will be added later. Thank you @lmars for suggestions on implementation!

File: swarm/api/http/server.go (hunk 1 above)
Comment (zelig) [reply swarm/api/http/server.go:986]: a year ago i had an issue to package requests but i dont think it is worth it. more likely it is the received chunk deliveries is what clogs incoming processing. we should look into this more thoroughly with metrics after chunk reffactor

File: swarm/api/http/server.go [hunk 2]
```diff
@@ -975,7 +976,38 @@ func (s *Server) HandleGetFile(w http.ResponseWriter, r *Request) {
...
+const getFileBufferSize = 4 * 32 * 1024
```
Comment (zelig) [inline swarm/api/http/server.go:987]: this constant looks suspiciously low to me. I think your experiments were influenced by the overall number of peers, if there are more peers in a row (parctically in PO 0) then  i bet we can afford a higher readahead buffer. 
Lets leave it as is and make sure we experiment with this after chunk refactor

**Review by zelig — APPROVED** [review]

File: swarm/api/http/server.go (hunk 2 above)
Comment (janos) [reply swarm/api/http/server.go:987]: Thanks Viktor. This surely needs more experimenting and testing.

**Review by nonsense — APPROVED** [review]

**Comment by janos** [comment]
Thank you for reviews.

**Pull request merged**